File: kotlin/src/main/kotlin/io/github/ocirne/aoc/year2024/Day2.kt
package io.github.ocirne.aoc.year2024

import io.github.ocirne.aoc.AocChallenge
import kotlin.math.abs

fun List<Int>.withoutIndex(index: Int): List<Int> = this.subList(0, index)  + this.subList(index + 1, this.size)

class Day2(val lines: List<String>) : AocChallenge(2024, 2) {

    private fun isSafeDeltas(deltas: List<Int>): Boolean {
        return (deltas.all { it > 0 } || deltas.all { it < 0 }) && deltas.all { abs(it) < 4}
    }

    private fun isSafe(line: String, withDampener: Boolean = false): Boolean {
        val values = line.split(' ').map(String::toInt)
        val deltas = values.zipWithNext().map { (x, y) -> y - x }
        if (isSafeDeltas(deltas)) {
            return true
        }
        if (!withDampener) {
            return false
        }
        for (i in values.indices) {
            val subDeltas = values.withoutIndex(i).zipWithNext().map { (x, y) -> y - x }
            if (isSafeDeltas(subDeltas)) {
                return true
            }
        }
        return false
    }

    override fun part1(): Int {
        return lines.count(::isSafe)
    }

    override fun part2(): Int {
        return lines.count { line -> isSafe(line, withDampener = true) }
    }
}

File: python/src/aoc/year2024/day2.py
from aoc.util import load_input, load_example


def check_part1(t):
    d = [a - b for a, b in zip(t[:-1], t[1:])]
    if not (all(x > 0 for x in d) or all(x < 0 for x in d)):
        return False
    return all(1 <= abs(x) <= 3 for x in d)


def check_part2(t):
    if check_part1(t):
        return True
    return any(check_part1(t[:i] + t[i + 1 :]) for i in range(len(t)))


def count_safe_reports(lines, f):
    return sum(f(list(map(int, line.split()))) for line in lines)


def part1(lines):
    """
    >>> part1(load_example(__file__, "2"))
    2
    """
    return count_safe_reports(lines, check_part1)


def part2(lines):
    """
    >>> part2(load_example(__file__, "2"))
    4
    """
    return count_safe_reports(lines, check_part2)


if __name__ == "__main__":
    data = load_input(__file__, 2024, "2")
    print(part1(data))
    print(part2(data))
